plugin/sampling/strategystore/static/strategy_store.go

```diff
@@ -29,31 +32,103 @@ import (
 )
 
 type strategyStore struct {
+	// to allow concurrent update of sampling strategies
+	lock sync.RWMutex
+
 	logger *zap.Logger
 
 	defaultStrategy   *sampling.SamplingStrategyResponse
 	serviceStrategies map[string]*sampling.SamplingStrategyResponse
 }
 
+func (h *strategyStore) loadAndParseStrategies(strategiesFile string) error {
+	s, err := loadStrategies(strategiesFile)
+	if err != nil {
+		h.logger.Warn("Using the last saved configuration for sampling strategies.", zap.Error(err))
+		return err
+	}
+
+	h.logger.Info("Updating sampling strategies file!", zap.Any("Strategies", s))
+	h.parseStrategies(s)
+	return nil
+}
+
 // NewStrategyStore creates a strategy store that holds static sampling strategies.
 func NewStrategyStore(options Options, logger *zap.Logger) (ss.StrategyStore, error) {
 	h := &strategyStore{
 		logger:            logger,
 		serviceStrategies: make(map[string]*sampling.SamplingStrategyResponse),
 	}
-	strategies, err := loadStrategies(options.StrategiesFile)
-	if err != nil {
+
+	// Read strategies
+	if err := h.loadAndParseStrategies(options.StrategiesFile); err != nil {
 		return nil, err
 	}
-	h.parseStrategies(strategies)
+
+	// Watch strategies file for changes.
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		logger.Error("failed to create a new watcher for the sampling strategies file", zap.Error(err))
+	}
+
+	go h.runWatcherLoop(watcher, options.StrategiesFile)
+
+	if err = watcher.Add(options.StrategiesFile); err != nil {
+		logger.Error("error adding watcher to file", zap.String("file", options.StrategiesFile), zap.Error(err))
+	} else {
+		logger.Info("watching", zap.String("file", options.StrategiesFile))
+	}
+
+	dir := filepath.Dir(options.StrategiesFile)
+	err = watcher.Add(dir)
+	if err != nil {
+		h.logger.Error("error adding watcher to dir", zap.String("dir", dir), zap.Error(err))
+	} else {
+		h.logger.Info("watching", zap.String("dir", dir))
+	}
+
 	return h, nil
 }
 
+func (h *strategyStore) runWatcherLoop(watcher *fsnotify.Watcher, strategiesFile string) {
+	for {
+		select {
+		case event := <-watcher.Events:
+			if event.Op&fsnotify.Remove == fsnotify.Remove {
+				if event.Name == strategiesFile {
+					h.logger.Warn("the sampling strategies file has been removed")
+
+					// This is a workaround for k8s configmaps. Since k8s loads configmaps as
+					// symlinked files within the containers, changes to the configmap register
+					// as `fsnotify.Remove` events.
+					if err := watcher.Add(strategiesFile); err != nil {
```

Comment:
Isn't there a race condition here? If the file was removed, even temporarily, and you call `watcher.Add()` on non-existing file, will it not error out immediately?

Reply:
So this line is specifically added for the k8s configmap scenario. It is added to handle successive reloads without restarting the collector (without this line, the collector reloads the strategies only the first time the file was changed). I've verified that it works. 

However, if the config file was manually removed, yes it will error out immediately, but that works for us.